bowler/tests/lib.py

```diff
@@ -0,0 +1,180 @@
+        return msg
+
+    def run_bowler_modifier(
+        self,
+        input_text,
+        selector=None,
+        modifier=None,
+        selector_func=None,
+        modifier_func=None,
+        in_process=True,
+    ):
+        """Returns the modified text."""
+
+        if not (selector or selector_func):
+            raise ValueError("Pass selector")
+        if not (modifier or modifier_func):
+            raise ValueError("Pass modifier")
+
+        exception_queue = multiprocessing.Queue()
+
+        def local_modifier(node, capture, filename):
+            # When in_process=False, this runs in another process.  See notes below.
+            try:
+                modifier(node, capture, filename)
+            except Exception as e:
+                exception_queue.put(e)
+
+        with tempfile.NamedTemporaryFile(suffix=".py") as f:
+            # TODO: I'm almost certain this will not work on Windows, since
+            # NamedTemporaryFile has it already open for writing.  Consider
+            # using mktemp directly?
+            with open(f.name, "w") as fw:
+                fw.write(input_text + "\n")
+
+            if selector_func:
+                query = selector_func([f.name])
+            else:
+                query = Query([f.name]).select(selector)
+
+            if modifier_func:
+                # N.b. exceptions may not work
+                query = modifier_func(query)
+            else:
+                query = query.modify(local_modifier)
+
+            # We require the in_process parameter in order to record coverage properly,
+            # but it also helps in bubbling exceptions and letting tests read state set
+            # by modifiers.
+            query.execute(
+                interactive=False, write=True, silent=False, in_process=in_process
+            )
+
+            # In the case of in_process=False (mirroring normal use of the tool) we use
+            # the queue to ship back exceptions from local_process, which can actually
+            # fail the test.  Normally exceptions in modifiers are not printed unless
+            # you pass --debug.
+            if not exception_queue.empty():
+                raise AssertionError from exception_queue.get()
+
+            with open(f.name, "r") as fr:
+                return fr.read().rstrip()
+
+    def run_bowler_modifiers(self, cases, selector=None, modifier=None):
+        for input, expected in cases:
+            with self.subTest(input):
+                output = self.run_bowler_modifier(input, selector, modifier)
+                self.assertEqual(expected, output)
+
+    def parse_line(self, source: str) -> LN:
+        grammar = pygram.python_grammar_no_print_statement
+        driver = Driver(grammar, convert=pytree.convert)
+        # Skip file_input, simple_stmt
+        return driver.parse_string(source + "\n").children[0].children[0]
+
+
+class BowlerTestCaseTest(BowlerTestCase):
```

Comment:
maybe put this somewhere besides test/lib.py?

Reply:
It's the meta-tests that ensure the BowlerTestCase itself works :)